posts/2021/01/enhancements-to-the-guvectorize-decorator.md

```diff
@@ -0,0 +1,150 @@
+
+```python
+>>> x, y = np.arange(10, dtype=np.float), 10
+>>> res = np.zeros_like(x)
+>>> guvec(x, y, res)
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+TypeError: ufunc 'guvec' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
+```
+
+## Dynamic compilation
+
+In Numba 0.53, one can omit the first argument to build a dynamic generalized universal function. For instance, consider the function `dyn_gufunc` below:
+
+```python
+@guvectorize('(n),()->(n)')
+def dyn_guvec(x, y, res):
+    for i in range(x.shape[0]):
+        res[i] = x[i] + y
+
+>>> dyn_guvec
+<numba._GUFunc 'dyn_guvec'>
+>>> dyn_guvec.ufunc.types
+[]
+```
+
+As one makes calls to `dyn_guvec`, new kernels will be generated for previously unsupported input types. The following set of interactions will illustrate how dynamic compilation works for a dynamic generalized ufunc:
+
+```python
+>>> x = np.arange(10).reshape(5, 2)
+>>> y = 10
+>>> res = np.zeros_like(x)
+>>> dyn_guvec(x, y, res)
+>>> res
+array([[10, 11],
+       [12, 13],
+       [14, 15],
+       [16, 17],
+       [18, 19]])
+>>> dyn_guvec.types
+['ll->l']
+```
+
+If this was a normal guvectorize function, one would have seen an exception complaining that the gufunc could not handle the given input. One can add additional loops by calling `dyn_guvec` with new types:
+
+```python
+>>> x_f = np.arange(5, dtype=np.float)
+>>> y_f = 10.0
+>>> res_f = np.zeros_like(x_f)
+>>> dyn_guvec(x_f, y_f, res_f)
+>>> dyn_guvec.types  # shorthand for dyn_guvec.ufunc.types
+['ll->l', 'dd->d']
+```
+
+## Current limitations
+
+In NumPy, it is fine to omit the output argument when calling a generalized ufunc.
+
+```python
+>>> a = np.arange(10).reshape(5, 2)
+>>> b = 10
+>>> guvec(a, b)
+array([[10, 11],
+       [12, 13],
+       [14, 15],
+       [16, 17],
+       [18, 19]])
+```
+
+The same is not possible in a dynamic ufunc. Numba would have to guess the output type and shape to correctly generate code based on the input and signature.
+
+```python
+>>> dyn_guvec(a, b)
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+  File "path/to/numba/np/ufunc/gufunc.py", line 134, in __call__
+    raise TypeError(msg)
+TypeError: Too few arguments for function 'dyn_guvec'. Note that the pattern `out = gufunc(Arg1, Arg2, ..., ArgN)` is not allowed. Use `gufunc(Arg1, Arg2, ..., ArgN, out) instead.
+```
+
+## Next steps
+
+In the future we would like to bring `@guvectorize` capabilities closer to the `@vectorize` one. For instance, currently it is not possible to call a guvectorize function from a jitted (`@jit`) function. Some work needs to be done towards this direction.
+
+We would like to thank the [D.E. Shaw group](https://www.deshaw.com/) for sponsoring this work.
```

Comment:
Presumably the wording of this blurb should be signed off by DEShaw, if it wasn't already. This is worded differently from other blurbs we used in the past.

Reply:
Got a new one from Eric and Troy